setup.py

```diff
@@ -519,21 +540,229 @@ def add_dir_name(dir_name, filenames):
+            warnings.warn(txt.format(swig_ver_str))
+
+            return 0, 0, 0
+
+        return tuple(int(x) for x in m.groups()[0:3])
+
     def run(self):
+        swig_ver = self.extract_swig_version(
+            str(subprocess.check_output(['swig', '-version'])))
+
+        if swig_ver < self.swig_min_ver:
+            txt = 'SWIG version {0}.{1}.{2} ' \
+                  'lower than the required version >= {3}.{4}.{5}. ' \
+                  'This will likely cause build errors!'
+
+            warnings.warn(txt.format(*(swig_ver + self.swig_min_ver)))
+
         self.run_command('build_ext')
         build.run(self)
 
 
 class CustomInstall(install):
     def run(self):
         self.run_command('build_ext')
-        self.do_egg_install()
+        install.run(self)
+
+
+class TickCommand(Command):
+    tick_dir = os.path.abspath(os.path.join(os.curdir, 'tick'))
+    cpp_build_dir = os.path.abspath(os.path.join(build_dir, 'cpptest'))
+
+    user_options = []
+
+    def initialize_options(self):
+        """Set default values for options."""
+        pass
+
+    def finalize_options(self):
+        """Post-process options."""
+        pass
+
+
+class BuildRunCPPTests(TickCommand):
+    description = 'build and run tick C++ tests'
+
+    def run(self):
+        self.run_command('makecpptest')
+        self.run_command('runcpptest')
+
+
+class RunCPPTests(TickCommand):
+    description = 'run tick C++ tests'
+
+    def run(self):
+        make_cmd = ['make', 'check']
+        subprocess.check_call(make_cmd, cwd=self.cpp_build_dir)
+
+
+class BuildCPPTests(TickCommand):
+    build_jobs = 1
+    description = 'build tick C++ tests'
+    user_options = [
+        ('build-jobs=', 'j',
+         'number of parallel build jobs (default is number of available CPU '
+         'cores reported by Python)'),
+    ]
+
+    def initialize_options(self):
+        """Set default values for options."""
+        self.build_jobs = multiprocessing.cpu_count()
+
+    def run(self):
+        relpath = os.path.relpath(self.tick_dir, self.cpp_build_dir)
+        cmake_exe = os.environ.get('TICK_CMAKE', 'cmake')
+
+        from distutils import sysconfig
+
+        cmake_cmd = [cmake_exe,
+                     '-DCMAKE_BUILD_TYPE=Debug',
+                     '-DPYTHON_INCLUDE_DIR={}'.format(sysconfig.get_python_inc()),
+                     relpath]
+
+        # Feed the path to the built C++ extensions so CMake does not have to
+        # build them again
+        for mod in tick_modules:
+            full_path = os.path.abspath(os.path.join(mod.module_ref.build, mod.module_ref.lib_filename))
+
+            cmake_cmd.append(
+                '-DTICKLIB_{}={}'.format(mod.ext_name.upper(),
+                                            full_path))
+
+        os.makedirs(os.path.join(self.cpp_build_dir, 'cpptest'), exist_ok=True)
+        subprocess.check_call(cmake_cmd, cwd=self.cpp_build_dir)
+
+        make_cmd = ['make', 'all', '-j{}'.format(self.build_jobs)]
```

Comment:
Even if that's not likely to happen, does this parallel build imply that test modules cannot have cross dependencies?